heartbeat/monitors/wrappers/monitorstate/tracker.go

```diff
@@ -0,0 +1,106 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package monitorstate
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/v7/heartbeat/monitors/stdfields"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// NewTracker tracks state across job runs. It takes an optional
+// state loader, which will try to fetch the last known state for a never
+// before seen monitor, which usually means using ES. If set to nil
+// it will use ES if configured, otherwise it will only track state from
+// memory.
+func NewTracker(sl StateLoader, flappingEnabled bool) *Tracker {
+	if sl == nil {
+		sl = NilStateLoader
+	}
+	return &Tracker{
+		states:          map[string]*State{},
+		mtx:             sync.Mutex{},
+		stateLoader:     sl,
+		flappingEnabled: flappingEnabled,
+	}
+}
+
+type Tracker struct {
+	states          map[string]*State
+	mtx             sync.Mutex
+	stateLoader     StateLoader
+	flappingEnabled bool
+}
+
+// StateLoader has signature as loadLastESState, useful for test mocking, and maybe for a future impl
+// other than ES if necessary
+type StateLoader func(stdfields.StdMonitorFields) (*State, error)
+
+func (t *Tracker) RecordStatus(sf stdfields.StdMonitorFields, newStatus StateStatus) (ms *State) {
+	//note: the return values have no concurrency controls, they may be unsafely read unless
+	//copied to the stack, copying the structs before  returning
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	state := t.getCurrentState(sf)
+	if state == nil {
+		state = newMonitorState(sf, newStatus, 0, t.flappingEnabled)
+		t.states[sf.ID] = state
+	} else {
+		state.recordCheck(sf, newStatus)
+	}
+	// return a copy since the state itself is a pointer that is frequently mutated
+	return state.copy()
+}
+
+func (t *Tracker) getCurrentState(sf stdfields.StdMonitorFields) (state *State) {
+	if state, ok := t.states[sf.ID]; ok {
+		return state
+	}
+
+	tries := 3
```

Comment:
I think this would be fine for `run_once` mode because it should be constrained to `1^2+2^2+500*3=6500ms` total delay + the time to actually exec the requests

Reply:
To add a little more detail here, the internet can be a little flakey sometimes, so I think it's a good thing to do a sort of minimal retry along the lines we do here for `run_once` as well. Also, the regular case is similar to `run_once` in that we don't want to impede the startup of heartbeat if ES is down, but it's OK to delay it... a little bit. So, it would make sense to use the same logic for both.